lib/src/local_working_copy.rs

```diff
@@ -640,47 +641,109 @@ fn can_create_new_file(disk_path: &Path) -> Result<bool, CheckoutError> {
             }
         },
     };
-    reject_reserved_existing_path(disk_path).inspect_err(|_| {
-        if new_file_created {
-            fs::remove_file(disk_path).ok();
-        }
-    })?;
-    if new_file_created {
+
+    let new_file_created = new_file.is_some();
+
+    if let Some(new_file) = new_file {
+        reject_reserved_existing_file(new_file, disk_path).inspect_err(|_| {
+            // We keep the error from `reject_reserved_existing_file`
+            let _ = fs::remove_file(disk_path);
+        })?;
+
         fs::remove_file(disk_path).map_err(|err| CheckoutError::Other {
             message: format!("Failed to remove temporary file {}", disk_path.display()),
             err: err.into(),
         })?;
+    } else {
+        reject_reserved_existing_path(disk_path)?;
     }
+
     Ok(new_file_created)
 }
 
 const RESERVED_DIR_NAMES: &[&str] = &[".git", ".jj"];
 
+/// Maps the result from constructing a `same_file::Handle` so that
+/// we get either an `Option<same_file::Handle>` which may be `None`
+/// if the file doesn't exist, or a `CheckoutError` indicating the
+/// error encountered when attempting to construct the `Handle`
+fn map_samefile_result(
+    disk_path: &Path,
+    result: std::io::Result<same_file::Handle>,
+) -> Result<Option<same_file::Handle>, CheckoutError> {
+    match result {
+        Ok(val) => Ok(Some(val)),
+        // If the existing disk_path pointed to the reserved path, the
+        // reserved path would exist.
+        Err(err) if err.kind() == io::ErrorKind::NotFound => Ok(None),
+        Err(err) => Err(CheckoutError::Other {
+            message: format!("Failed to validate path {}", disk_path.display()),
+            err: err.into(),
+        }),
+    }
+}
```

Comment:
Maybe we can instead add `same_file_handle_from_file()`/`same_file_handle_from_path()` helpers? Since "from_file" shouldn't fail with `NotFound` error, we can make it a bit stricter.

Reply:
Let's continue this discussion on https://github.com/jj-vcs/jj/pull/7742